contracts/src/libs/BlockNumbers.sol

```diff
@@ -0,0 +1,171 @@
+        return self.prefix & type(uint64).max;
+    }
+
+    /**
+     * @notice Create an iterator over the block numbers.
+     * @dev The iterator is implemented to minimize storage reads.
+     */
+    function iter(T storage self) internal view returns (Iterator memory it) {
+        // solhint-disable-next-line no-inline-assembly
+        assembly ("memory-safe") {
+            mstore(it, self.slot)
+        }
+        unchecked {
+            it.data = self.prefix;
+            // TODO(nlordell): If we _really_ want to optimize, we can pack `cnt`, `rem` and
+            // `first` into a single slot, but that isn't necessary for now.
+            it.cnt = it.data & type(uint64).max;
+            it.rem = 3;
+            it.first = true;
+        }
+    }
+
+    /**
+     * @notice Returns the count of remaining items in the iterator.
+     */
+    function count(
+        Iterator memory self
+    ) internal pure returns (uint256 length) {
+        return self.cnt;
+    }
+
+    /**
+     * @notice Moves the iterator to the next block number.
+     */
+    function next(Iterator memory self) internal view returns (bool remaining) {
+        remaining = self.cnt != 0;
+        skip(self, 1);
+    }
+
+    /**
+     * @notice Skips items in the iterator.
+     */
+    function skip(Iterator memory self, uint256 n) internal view {
+        if (self.cnt < n) {
+            self.cnt = 0;
+            return;
+        }
+
+        unchecked {
+            self.cnt -= n;
+            if (self.rem >= n) {
+                self.data = self.data >> (n << 6);
+                self.rem -= n;
+            } else {
+                n -= self.rem;
+                uint256 slots = n >> 2;
+                if (self.first) {
+                    // self.slot = keccak256(abi.encode(self.slot)) + slots;
+                    // solhint-disable-next-line no-inline-assembly
+                    assembly ("memory-safe") {
+                        mstore(self, add(keccak256(self, 32), slots))
+                    }
+                    self.first = false;
+                } else {
+                    self.slot += 1 + slots;
+                }
+                self.rem = 4 - (n & 3);
+                uint256 shift = (3 - self.rem) << 6;
+                // self.data = sload(self.slot) >> shift
+                // solhint-disable-next-line no-inline-assembly
+                assembly ("memory-safe") {
+                    mstore(add(self, 32), shr(shift, sload(mload(self))))
+                }
+            }
+        }
+    }
+
+    /**
+     * @notice Truncates the iterator to a maximum of `cnt` items.
+     */
+    function take(Iterator memory self, uint256 cnt) internal pure {
+        if (self.cnt > cnt) {
+            self.cnt = cnt;
+        }
+    }
+
+    /**
+     * @notice Returns the current value of the iterator.
+     * @dev The caller must only call this function after a call to `next` which returned `true`.
```

Comment:
Can upper case "must only", as this is important as this might return unexpected behavior in some cases if `skip` was called before? I.e. if the number of items skipped is greater than the number of remaining items.